---
knit: (function(input_file, encoding) {
  out_dir <- '.';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'LineagesAndImport.html'))})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

# Import and local transmission analysis


```{r, include=FALSE}
#Load required packages and local variables
library(ape)
library(LineageHomology)
library(skygrowth)
library(grid)
library(scales) #Make more axis ticks.
library(gridExtra)
library(ggplot2)
library(phylodyn)
library(pbapply)

#Notes:
#Time consuming sections are commented out. The results are saved and loaded on new runs instead of rerunning every time. 
#Had to install an older version of RcppArmadillo to get the script running
#"install.packages("https://cran.r-project.org/src/contrib/Archive/RcppArmadillo/RcppArmadillo_0.9.900.3.0.tar.gz", repos=NULL, type="source")"

#Setup paths
path_main = "~/Dropbox/10000_Lovers/Phylogeography"
path_data = "~/Dropbox/10000_Lovers/01_LineageHomology/11_saved_data/"
path_results = "~/Dropbox/10000_Lovers/01_LineageHomology/12_saved_results/"

source("~/Dropbox/10000_Lovers/01_LineageHomology/sourcefunctions/environment_variables.R")
```

The input for this analysis is the results from the maximum likelihood geographical state mapping on the full phylogenetic tree. 

```{r, include=FALSE}
#Names of the norwegian sequences
nor_tips = names(locations_NOR[which(locations_NOR=="Norway")])
aus_tips = names(locations_NOR[which(locations_AUS=="Australia")])
usa_tips = names(locations_NOR[which(locations_USA=="A_USA")])

#Replicate transmission lineage (TL) counts.
# replicates_NOR = pbreplicate(
#   5,
#   LineageHomology::LineageHomology_w_uncertainty_v2(
#     tree=tree,
#     ace_nodes=ace_NOR$lik.anc,
#     ace_tips = to.matrix(locations_NOR, seq=c("Norway", "RoW")),
#     start_time = 1750,give_tips = nor_tips
#   )
# )

#Australia
# replicates_AUS = pbreplicate(
#   5,
#   LineageHomology::LineageHomology_w_uncertainty_v2(
#     tree=tree,
#     ace_nodes=ace_AUS$lik.anc,
#     ace_tips = to.matrix(locations_AUS, seq=c("Australia", "RoW")),
#     start_time = 1750,give_tips = aus_tips
#   )
# )

#USA
# replicates_USA = pbreplicate(
#   5,
#   LineageHomology::LineageHomology_w_uncertainty_v2(
#     tree=tree,
#     ace_nodes=ace_USA$lik.anc,
#     ace_tips = to.matrix(locations_USA, seq=c("A_USA", "RoW")),
#     start_time = 1750,give_tips = usa_tips
#   )
# )
#save(replicates_NOR, file=paste0(path_results,"replicates_NOR.Rdata"))
#save(replicates_AUS, file=paste0(path_results,"replicates_AUS.Rdata"))
#save(replicates_USA, file=paste0(path_results,"replicates_USA.Rdata"))
load(file=paste0(path_results,"replicates_NOR.Rdata"))
load(file=paste0(path_results,"replicates_AUS.Rdata"))
load(file=paste0(path_results,"replicates_USA.Rdata"))
```

## Import versus local transmission for Norway, Australia and USA

From the mapped geographical lineage distributions we can derive the number of importation and local transmission events for each location (based on the available sequences). NB. we do not account for unobserved infections (sequences), and hence the estimated number only reflects the fraction of *sequenced* genomes that is assigned to specific lineages. The relative growth between the lineages (growth of an transmission lineage implies local transmission), is an important quantity and tells us how dominant specific lineages have become over time. This too could be affected by nonrandom sampling such as targeted sampling of lineages with anitibiotic resistance. 



```{r, include=TRUE}
LineageHomology::Summarize_import_local_transmission(replicates_NOR)
LineageHomology::Summarize_import_local_transmission(replicates_AUS)
LineageHomology::Summarize_import_local_transmission(replicates_USA)

```


In Norway we estimate that there is 29% importation. This is more than twice the estimated amount in USA and Australia (each ca. 13%)

## Cumulative growth of transmission lineages

The following plots show the cumulative number of sequences obtained from each estimated transmission lineage over time
```{r, include=TRUE, warning=FALSE}
name_date_NOR = name_date[name_date$name%in%nor_tips,]
linfo_NOR = lineage_info(Result_NOR,name_date = name_date_NOR)
LineageHomology::lineage_growth_cumulative(linfo_NOR,datelims=c("2015-01-01","2019-06-15","1 year"))


name_date_AUS = name_date[name_date$name%in%aus_tips,]
linfo_AUS = lineage_info(Result_AUS,name_date = name_date_AUS)
LineageHomology::lineage_growth_cumulative(linfo_AUS,datelims=c("2016-07-01","2018-06-15","5 months"))

name_date_USA = name_date[name_date$name%in%usa_tips,]
linfo_USA = lineage_info(Result_USA,name_date = name_date_USA)
LineageHomology::lineage_growth_cumulative(linfo_USA,datelims=c("2008-07-01","2019-06-15","2 year"))

```


## Import and local transmission over time

```{r remedy001, echo=F, include=F}
#Extract importation and local transmission on monthly basis
# extract_indices = function(counts,time_interval=c("2000-05-01","2016-05-01")){
#   clims = decimal_date(as.Date(time_interval))
#   indselect=which(counts$Week_time>=clims[1]&counts$Week_time<=clims[2])
#   counts$Import=counts$Import[indselect]
#   counts$LC=counts$LC[indselect]
#   counts$Week_time=counts$Week_time[indselect]
# }

source("~/Dropbox/Github/10000_Ngon_genomes/R/plot_import_local.R")


```

```{r remedy002, echo=F, include =F}
# imploc_NOR = LineageHomology::import_local_transmission(tree,replicates_NOR,start_time=1750,time_interval_size = 1)
# save(imploc_NOR, file=paste0(path_results,"imploc_NOR.Rdata"))
load(file=paste0(path_results,"imploc_NOR.Rdata"))

# imploc_AUS = LineageHomology::import_local_transmission(tree,replicates_AUS,start_time=1750,time_interval_size = 1)
# save(imploc_AUS, file=paste0(path_results,"imploc_AUS.Rdata"))
load(file=paste0(path_results,"imploc_AUS.Rdata"))

# imploc_USA = LineageHomology::import_local_transmission(tree,replicates_USA,start_time=1750,time_interval_size = 1)
# save(imploc_USA, file=paste0(path_results,"imploc_USA.Rdata"))
load(file=paste0(path_results,"imploc_USA.Rdata"))

```

```{r remedy003, echo=F}
plot_importation_local(tree = tree,result_import_local_transmission = imploc_NOR,start_time = 1750,date_breaks = "2 years",time_interval = c("2000-01-01","2018-01-01"), main_title = "Norway")
#dev.off()
```

```{r remedy004, echo=F}
plot_importation_local(tree = tree,result_import_local_transmission = imploc_USA,start_time = 1750,date_breaks = "2 years",time_interval = c("2000-01-01","2018-01-01"), main_title = "USA")
#dev.off()
```

```{r remedy005, echo=F}
g1 = plot_importation_local(tree = tree,result_import_local_transmission = imploc_AUS,start_time = 1750,date_breaks = "2 years",time_interval = c("2000-01-01","2018-01-01"), main_title = "Australia")
#dev.off()
```

