---
title: "Downsampling analysis"
output:
    html_document:
        toc: true
        toc_float: true
        theme: united
        page-layout: full
 #    knit: (function(input_file, encoding) {
 #  out_dir <- '.';
 #  rmarkdown::render(input_file,
 # encoding=encoding,
 # output_file=file.path(dirname(input_file), out_dir, 'global_analysis.html'))})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, warning=FALSE, collapse=T, cache=T, 'styler', tidy.opts=list(strict=T))

```

## Outline

In this analysis we evaluate the performance and. stability of the estimates of LineageHomology in various scenarios of sampling density.

Our main question is: Do we have enough genomes to reliably estimates the relative values of import, export and local transmission.

To answer this question we perform a down-sampling analysis by removing isolates from the focal location, and isolates from outside of the locations. The results in our main analysis is based on a binary division of the isolates to the focal location: Norway, Victoria, Europe or the USA, and the rest of the rest-of-the-world (ROW). For example, when the focus is shifted from Norway from Victoria, all the genomes from Norway will be placed in the ROW category, and vice versa.

For each focal location we down-sample the focal location and the rest of the world in a step-wise fashion. In the case of the Norwegian analysis, there are $1723$ genomes from Norway. Fom the phylogeny we step-wise remove isolates $0, 100, 200,..., 1690, 1700,1710$ and trim the terminal and possibly internal branches. For each down-sampled dataset we re-estimate the geography by maximum likelihood using ace and rerun LineageHomology to estimate import, export and local transmission. In the Norwegian analysis, there are $8007$ genomes from ROW, and we step-wise remove $0, 1000, 2000, ... , 7700, 7800, 7900, 7950$, and similarly re-estimate.

```{r}
#Load needed packages
library(LineageHomology)
library(BactDating)
library(phytools)
library(ape)
library(geiger)
library(ggplot2)
library(stringr)
library(dplyr)
library(scales)
```

## Setting up the downsampling functions

```{r}
# Load full result
load("/Users/magnusnygardosnes/Dropbox/10000_Lovers/01_LineageHomology/11_saved_data/data_for_downsampling.Rdata")

#Set seed


# Downsample
downsample = function(tree, locations, downsample_location = "ROW", remove_tips=1000) {
  if(remove_tips!=0){
    indexes = which(locations==downsample_location)
    ind_select = sample(indexes,remove_tips,replace = F)
    select_tips = names(locations[ind_select])
    downsampled_tree = ape::drop.tip(tree,tip =select_tips,trim.internal = T)
    downsampled_locations = locations[-ind_select]  
  }
  else{
    downsampled_locations = locations
    downsampled_tree=tree
  }
  
  #Reorder downsampled locations after tips
  reorder_downsample = match(downsampled_tree$tip.label, names(downsampled_locations))
  downsampled_locations=downsampled_locations[reorder_downsample]
  return(list(downsampled_tree = downsampled_tree,
              downsampled_locations = downsampled_locations))
}


#covariance_matrix = vcvPhylo(tree2)

#NOT CURRENTLY IN USE
downsample_based_on_ancestral_relationship = function(tree, locations, downsample_location = "ROW", remove_tips=1000) {
  if(remove_tips!=0){
    
    ## NEED TO FILL IN THESE LINES.
    #indexes = which(locations==downsample_location)
    #ind_select = sample(indexes,remove_tips,replace = F)
    
    select_tips = names(locations[ind_select])
    downsampled_tree = ape::drop.tip(tree,tip =select_tips,trim.internal = T)
    downsampled_locations = locations[-ind_select]  
  }
  else{
    downsampled_locations = locations
    downsampled_tree=tree
  }
  
  #Reorder downsampled locations after tips
  reorder_downsample = match(downsampled_tree$tip.label, names(downsampled_locations))
  downsampled_locations=downsampled_locations[reorder_downsample]
  return(list(downsampled_tree = downsampled_tree,
              downsampled_locations = downsampled_locations))
}

#May need to reorder locations after downsampling
set.seed(1)
# Test
#downsampled_1000_NOR = downsample(tree2,locations_NOR,remove_tips=1000,downsample_location = "Norway")
#ace_downsampled_1000_NOR = ace(x=downsampled_1000_NOR$downsampled_locations, phy= downsampled_1000_NOR$downsampled_tree, type="discrete", mod="ARD")


downsample_and_calculate = function(tree, locations, remove_tips=1000, coded_location="Norway",downsample_location="ROW") {
  #DEBUG
  #tree=tree2;location=locations_NOR;remove_tips=7000;
  
  downsampled = downsample(tree,locations,remove_tips=remove_tips, downsample_location = downsample_location)
  ace_downsample = ace(x=downsampled$downsampled_locations, phy= downsampled$downsampled_tree, type="discrete", mod="ARD")
  tips = names(downsampled$downsampled_locations[downsampled$downsampled_locations==coded_location])
  LH= LineageHomology::LineageHomology_v2(tree=downsampled$downsampled_tree,
                                 give_tips =tips,
                                 ace_nodes = ace_downsample$lik.anc,
                                 ace_tips = to.matrix(downsampled$downsampled_locations,seq=c(coded_location,"ROW")),
                                 start_time = start_date)
  
  return(LH)
}

# Alternative plotting approaches.
# An alternative plotting method could be to use the moving averages
# The downsampling vector could only contain unique values, and so would explore more of the shape of the curve.
downsample_for_values = function(downsample_vector,reps=3,focal_location="Norway",downsample_row = T) {
  
  #Debug 
  # downsample_vector= c(0, 1000);reps=3;focal_location="Norway";downsample_row = T
  # downsample_vector = downsample_vector;reps = 1; focal_location = "Norway";downsample_row = F;
  # downsample_vector = remove_tips_values_AUS;
  # reps = 1;                      
  # focal_location = "Victoria";                      
  # downsample_row = T;
  
  result_list = list()
  
  if(downsample_row==T) {
    total_genomes_list = c("Norway"=8007,"Victoria"=7530,"Europe"=5628,"The USA"=7342)
    downsample_location="ROW"
    }
  
  
  
  locations_list  = list("Norway"=locations_NOR,"Victoria"=locations_AUS,"Europe"=locations_EUR,"The USA"=locations_USA)
  location_coding = c("Norway"="Norway","Victoria"="Australia","Europe"="Europe","The USA"="A_USA")
  
  if(downsample_row==F) {
    total_genomes_list = c("Norway"=1726,"Victoria"=2203,"Europe"=4105,"The USA"=2391)
    downsample_location = as.character(location_coding[focal_location])
  }
  # Obtain correct variables for each downsampling
  total_genomes = as.numeric(total_genomes_list[focal_location])
  location = locations_list[focal_location][[1]]
  coded_location = as.character(location_coding[focal_location])
  
  remove_tips_values = downsample_vector
  remove_tips_values = c(rep(remove_tips_values, each = reps))
  
  for (i in 1:length(remove_tips_values)) {
  cat('\r' ,round(i/length(remove_tips_values)*100,3),  "%")
  result_list[[paste0(i,"downsampled_", remove_tips_values[i])]] = downsample_and_calculate(tree2, 
                                                                                                   location = location,
                                                                                                   remove_tips = remove_tips_values[i],
                                                                                                   coded_location = coded_location,
                                                                                            downsample_location=downsample_location)
  }
  names(result_list) = paste0("downsampled_", remove_tips_values)
  
  # Create a plot of the values with x-labels given by the names
  import_export_local = lapply(result_list, function(x) x$Import_Export_LocalTransmission)
  total = unlist(lapply(import_export_local, "[[", 1))+unlist(lapply(import_export_local, "[[", 3))
  imports = unlist(lapply(import_export_local, "[[", 1))/total
  exports = unlist(lapply(import_export_local, "[[", 2))/total
  local_transmission = unlist(lapply(import_export_local, "[[", 3))/total
  genomes_from_row = total_genomes-remove_tips_values
  
  return(list("genomes_from_row"=genomes_from_row,
              "total_genomes"=total,
              "imports"=imports,
              "exports"=exports,
              "local_transmission"=local_transmission,
              "focal_location"=focal_location,
              "downsample_row"=downsample_row))
}

plot_repeated_downsampling = function(downsampling_result) {
  
  downsampling_result$genomes_from_row
  
  df <- data.frame(genomes_from_row = downsampling_result$genomes_from_row,
                 imports = downsampling_result$imports,
                 exports = downsampling_result$exports,
                 local_transmission = downsampling_result$local_transmission)

  median_data = df %>% group_by(genomes_from_row) %>% summarize(med_imp=median(imports),med_exp =median(exports), med_loc =median(local_transmission))
  
  xlab = paste0("Genomes from ", downsampling_result$focal_location)
  title = paste0("Downsampling ", downsampling_result$focal_location)
  
  # Overwrite if it we are downsampling ROW
  if(downsampling_result$downsample_row) {
    xlab = paste0("Genomes from ROW")
    title = paste0("Downsampling ROW keeping ", downsampling_result$focal_location, " constant")
  }
  
  alpha1 = 0.5
  g1 = ggplot(df, aes(x = genomes_from_row)) +
  geom_line(aes(x = genomes_from_row,y = med_imp,color = "Import"), size = 1.5,  data=median_data) +
  geom_line(aes(x = genomes_from_row,y = med_exp,color = "Export"), size = 1.5,  data=median_data)  +
  geom_line(aes(x = genomes_from_row,y = med_loc,color = "Local transmission"), size = 1.5,  data=median_data)  +
  geom_point(aes(y = imports, color  = "Import"), size = 2,alpha=alpha1) +
  geom_point(aes(y = exports, color = "Export"), size = 2,alpha=alpha1) +
  geom_point(aes(y = local_transmission, color = "Local transmission"), size = 2,,alpha=alpha1) +
  scale_y_continuous(limits=c(0,1), oob=rescale_none) +
  labs(title = title,
       x = xlab,
       y = "Fraction attributed to transmission mode",
       color = "Transmission mode") +
    theme_bw(base_size=20)+
  theme(legend.position = c(0.5, 0.5),
        legend.direction = "vertical",
        panel.border = element_blank(),
        axis.line = element_line(colour = "black",size = 1, linetype = 1),
        legend.title.align = 0.5,
        legend.background = element_blank(),
        plot.title = element_text(hjust = 0.5))+scale_color_manual(values = c("Import"="steelblue","Export"="firebrick", "Local transmission"="seagreen"))
    
  g1
}

# Number of downsamples
reps = 5
save_figures = T
```

## Downsampling wrt. Norway

```{r}
set.seed(1)

load(file="/Users/magnusnygardosnes/Dropbox/10000_Lovers/01_LineageHomology/12_saved_results/downsampling_analysis.Rdata")
#Downsampling from Norway

# max=7950
# downsample_vector = floor(sort(c(seq(0,max,length.out=18),seq((max-100),max,20)[-length(seq((max-100),max,20))])))
# downsample_NOR_ROW = downsample_for_values(downsample_vector = downsample_vector,
#                       reps = reps,
#                       focal_location = "Norway",
#                       downsample_row = T)
# Create the plot using ggplot
 # change this to F if you don't want to save
if (save_figures) {
  png(filename="~/Dropbox/10000_Lovers/99_Figures_for_paper/Supplementary/downsampling_norway.png",units="in", width=10,height = 10,res = 300)  
}

g1 = plot_repeated_downsampling(downsample_NOR_ROW)
g1
if (save_figures) {
  dev.off()
}


# max=1700
# downsample_vector = floor(sort(c(seq(0,max,length.out=18),seq((max-100),max,20)[-length(seq((max-100),max,20))])))
# downsample_NOR = downsample_for_values(downsample_vector = downsample_vector,
#                       reps = reps,
#                       focal_location = "Norway",
#                       downsample_row = F)
if (save_figures) {
  png(filename="~/Dropbox/10000_Lovers/99_Figures_for_paper/Supplementary/downsampling_norway_row.png",units="in", width=10,height = 10,res = 300)
}

f1 = plot_repeated_downsampling(downsample_NOR)
f1
if (save_figures) {
  dev.off()
}
```

## Downsampling wrt. Victoria, Australia

```{r}
#Downsampling Australia
# max=7500
# downsample_vector = floor(sort(c(seq(0,max,length.out=18),seq((max-100),max,20)[-length(seq((max-100),max,20))])))
# downsample_AUS_ROW = downsample_for_values(downsample_vector = downsample_vector,
#                       reps = reps,
#                       focal_location = "Victoria",
#                       downsample_row = T)
# Create the plot using ggplot
# 
if (save_figures) {
  png(filename="~/Dropbox/10000_Lovers/99_Figures_for_paper/Supplementary/downsampling_victoria.png",units="in", width=10,height = 10,res = 300)
}

g2 = plot_repeated_downsampling(downsample_AUS_ROW)
g2
if (save_figures) {
  dev.off()
}



# max=2193
# downsample_vector = floor(sort(c(seq(0,max,length.out=18),seq((max-100),max,20)[-length(seq((max-100),max,20))])))
# downsample_AUS = downsample_for_values(downsample_vector = downsample_vector,
#                       reps = reps,
#                       focal_location = "Victoria",
#                       downsample_row = F)
if (save_figures) {
  png(filename="~/Dropbox/10000_Lovers/99_Figures_for_paper/Supplementary/downsampling_victoria_row.png",units="in", width=10,height = 10,res = 300)
}

f2 = plot_repeated_downsampling(downsample_AUS)
f2
if (save_figures) {
  dev.off()
}


```

## Downsampling wrt. Europe

```{r}
#Downsampling Europe

# max=5600
# downsample_vector = floor(sort(c(seq(0,max,length.out=18),seq((max-100),max,20)[-length(seq((max-100),max,20))])))
# downsample_EUR_ROW = downsample_for_values(downsample_vector = downsample_vector,
#                       reps = reps,
#                       focal_location = "Europe",
#                       downsample_row = T)
# Create the plot using ggplot
if (save_figures) {
  png(filename="~/Dropbox/10000_Lovers/99_Figures_for_paper/Supplementary/downsampling_europe_row.png",units="in", width=10,height = 10,res = 300)
}

g3 = plot_repeated_downsampling(downsample_EUR_ROW)
g3
if (save_figures) {
  dev.off()
}


# max=4095
# downsample_vector = floor(sort(c(seq(0,max,length.out=18),seq((max-100),max,20)[-length(seq((max-100),max,20))])))
# downsample_EUR = downsample_for_values(downsample_vector = downsample_vector,
#                       reps = reps,
#                       focal_location = "Europe",
#                       downsample_row = F)

if (save_figures) {
  png(filename="~/Dropbox/10000_Lovers/99_Figures_for_paper/Supplementary/downsampling_europe.png",units="in", width=10,height = 10,res = 300)
}
f3 = plot_repeated_downsampling(downsample_EUR)
f3

if (save_figures) {
  dev.off()
}



```

## Downsampling wrt. USA

```{r}
#Downsampling the USA

# max=7300
# downsample_vector = floor(sort(c(seq(0,max,length.out=18),seq((max-100),max,20)[-length(seq((max-100),max,20))])))
# downsample_USA_ROW = downsample_for_values(downsample_vector = downsample_vector,
#                       reps = reps,
#                       focal_location = "The USA",
#                       downsample_row = T)
# Create the plot using ggplot

if (save_figures) {
  png(filename="~/Dropbox/10000_Lovers/99_Figures_for_paper/Supplementary/downsampling_USA_row.png",units="in", width=10,height = 10,res = 300)
}
g4 = plot_repeated_downsampling(downsample_USA_ROW)
g4
if (save_figures) {
  dev.off()
}

# max=2380
# downsample_vector = floor(sort(c(seq(0,max,length.out=18),seq((max-100),max,20)[-length(seq((max-100),max,20))])))
# downsample_USA = downsample_for_values(downsample_vector = downsample_vector,
#                       reps = reps,
#                       focal_location = "The USA",
#                       downsample_row = F)
if (save_figures) {
  png(filename="~/Dropbox/10000_Lovers/99_Figures_for_paper/Supplementary/downsampling_USA.png",units="in", width=10,height = 10,res = 300)
}

f4 = plot_repeated_downsampling(downsample_USA)
f4
if (save_figures) {
  dev.off()
}


# save(downsample_NOR_ROW,downsample_NOR,downsample_AUS_ROW,downsample_AUS,downsample_EUR_ROW, downsample_EUR,downsample_USA_ROW,downsample_USA,
#   file="/Users/magnusnygardosnes/Dropbox/10000_Lovers/01_LineageHomology/12_saved_results/downsampling_analysis.Rdata")
```



```{r}
library(ggplot2)
library(patchwork)
# Arrange plots side-by-side
(f1 | g1)
(f2 | g2)
(f3 | g3)
(f4 | g4)
```


```{r, include=F}
## Fit exponential curves to determine if they converge, and to what value

generate_x_y = function(result_list, row_or_focal="row"){

  genomes_from_focal = c("Australia"=7530,"A_USA"=7342,"Norway"=8007, "Europe"=5628)
  if(row_or_focal != "row") {
    genomes_from_focal = c("Australia"=2203,"A_USA"=2391,"Norway"=1726, "Europe"=4105)
  }
  loc = names(result_list[[1]]$lineage_state)[1]
  genomes = genomes_from_focal[loc]-as.numeric(unlist(str_extract_all(names(result_list), "[0-9]+")))
  import_export_local = lapply(result_list, function(x) x$Import_Export_LocalTransmission)
  
  
  if(row_or_focal=="row"){
  total = unlist(lapply(import_export_local, "[[", 1))+unlist(lapply(import_export_local, "[[", 3))
  imports = unlist(lapply(import_export_local, "[[", 1))/total
  exports = unlist(lapply(import_export_local, "[[", 2))/totallocal_transmission = unlist(lapply(import_export_local, "[[", 3))/total
  #Fit an exponential model using nls()  
  }
  
  if(row_or_focal=="row"){
    import_export_local = lapply(result_list, function(x) x$Import_Export_LocalTransmission)
    total = sum(import_export_local[[1]][1:2])
    imports = unlist(lapply(import_export_local, "[[", 1))/total
    exports = unlist(lapply(import_export_local, "[[", 3))/total
    local_transmission = unlist(lapply(import_export_local, "[[", 2))/total
  }
  library(propagate)
  dat = data.frame(genomes=genomes, imports=imports,exports=exports, local_transmission=local_transmission)
  head(dat)
  model_import <- nls(imports ~ a + b * exp(r * genomes),
                      data = dat,
                      start = list(a =0.25, b=-0.25, r = -1e-4))
  
  dev.off()
  plot(dat$genomes, dat$imports, col = "red")
  lines(dat$genomes, pred, col = "blue") # model predictions
  lines(dat$genomes, pred[, "fit"], col = "blue") # model predictions
  lines(dat$genomes, pred[, "lwr"], col = rgb(0, 0, 1, alpha = 0.5)) # lower bound of confidence interval
  lines(dat$genomes, pred[, "upr"], col = rgb(0, 0, 1, alpha = 0.5)) # upper bound of confidence interval
  #Access the best fit parameters
  a <- coef(model_import)[1]
  b <- coef(model_import)[2]
  r <- coef(model_import)[3]
  #Calculate the limit as x approaches infinity
  limit <- a + b * exp(r * 20000)
}

# Convergence analysis

```




 <!-- ## Simulate phylogeny
# Discarded sections
```{r}
#1. Simulate phylogeny
set.seed(2)
dated_tree = BactDating::simdatedtree(nsam = 200,dateroot = 2000)
plot(ladderize(dated_tree))
axisPhylo(side=1, backward = F, root.time=2000)
```

## Simulate geography

We set up a transition rate matrix where the instaneous rate of moving to an different location is 1/10

|                          | **Transition Rate Matrix** |       |
|--------------------------|----------------------------|-------|
|                          | Location of interest       | Other |
| **Location of interest** | 0.9                        | 0.1   |
| **Other**                | 0.1                        | 0.9   |

This means that we expect the amount of local transmission to be 90% and 10% to be importation.

```{r}
#2. Simulate geography
# Define the transition matrix
transition_matrix <- matrix(c(0.9, 0.1, 0.1, 0.9), nrow = 2)

# Define the states (locations)
# TBD: Instead of sampling we can probably place the tips based on the covariance of the samples. 
location2 = c(rep("RoW",40), rep("Norway",160))

covariance_matrix = vcvPhylo(dated_tree)

location2 = sample(location2)
names(location2)=dated_tree$tip.label

# Compute the state probabilities using the ace function from the ape package
sim = make.simmap(tree=dated_tree, x=location2, mod="ER")
c1 = summary(sim)
node_states = to.matrix(c1$states,seq= c("RoW","Norway"))
tip_states = to.matrix(location2,seq= c("RoW","Norway"))


#plot the simulated history and probabilities
plotSimmap(sim, fsize = 0.5, mar=c(0.2,0.2,0.2,0.2))
plot.phylo(dated_tree,edge.width = 3,label.offset = 0.15, mar=c(0.2,0.2,0.2,0.2))
nodelabels(pie=node_states,cex=0.3,piecol=c("red","black"))
tiplabels(pie=tip_states, cex=0.3,piecol=c("red","black"))

```

## Calculate the true number of transmission lineages, imports, exports and local transmission

```{r}
result = LineageHomology(tree = dated_tree, ace_nodes = node_states,tip_states, start_time = 2000)

```

## Downsample the tree

```{r}
#1. Downsample

```

## Ideas

-   High sampling density means high probability of capturing closely related lineages. Should sample from the covariance matrix.
-   Downsampling the true phylogeny

## GPTprompts

I have a time dated phylogeny.\
I need to simulate locations on the phylogeny in R.\
I want the rate of staying in the same location to be 90% and the rate of moving to a different location to be 10%.\
Can you help me with the R code to do this?

...

I need the format to match the output from ace, so that I have the probabilitiy of each state on each node. Can you process the result to achieve that? --> 
